office-365-feed/operations.py

```diff
@@ -0,0 +1,275 @@
+        for region in regions_list:
+            if service == 'All':
+                url = 'https://endpoints.office.com/{0}/{1}?ClientRequestId={2}'.format(endpoint, region, unique_id)
+            else:
+                url = 'https://endpoints.office.com/{0}/{1}?ServiceAreas={2}' \
+                      '&ClientRequestId={3}'.format(endpoint, region, service, unique_id)
+
+            if latest_version_list:
+                for dict_ in [x for x in latest_version_list if x["instance"] == region]:
+                    urls_list.append({'lastUpdated': dict_.get("latest"),
+                                      'region': region,
+                                      'feedURL': url,
+                                      })
+            else:
+                urls_list.append({
+                    'region': region,
+                    'feedURL': url
+                })
+
+        return urls_list
+    except Exception as Err:
+        logger.error(str(Err))
+        raise ConnectorError(str(Err))
+
+
+def make_api_call(params):
+    # Retrieves all entries from the feed.
+    try:
+        result = []
+        for feed_obj in params.get('urls_list'):
+            feed_url = feed_obj.get('feedURL', '')
+            try:
+                response = requests.get(
+                    url=feed_url,
+                    verify=params.get('verify_ssl')
+                )
+                response.raise_for_status()
+                data = response.json()
+                if "latest" in data:
+                    result.append(data)
+                else:
+                    indicators = [i for i in data if 'ips' in i or 'urls' in i]  # filter empty entries
+                    result.extend(indicators)
+            except requests.exceptions.SSLError:
+                logger.error('An SSL error occurred.')
+                raise ConnectorError('An SSL error occurred.')
+            except requests.exceptions.ConnectionError:
+                logger.error('A connection error occurred.')
+                raise ConnectorError('A connection error occurred.')
+            except Exception as err:
+                logger.error(err)
+                raise ConnectorError(err)
+        return result
+    except Exception as Err:
+        logger.error(str(Err))
+        raise ConnectorError(str(Err))
+
+
+def get_latest_version(config):
+    try:
+        client_request_id = str(uuid.uuid4())
+        regions_list = convert_str_list(config.get('regions'))
+        params = {"urls_list": build_urls_dict(regions_list, services_list=["All"], unique_id=client_request_id,
+                                               endpoint="version"),
+                  "verify_ssl": config.get("verify_ssl")}
+
+        resp = make_api_call(params)
+        return resp
+    except Exception as Err:
+        logger.error(str(Err))
+        raise ConnectorError(str(Err))
+
+
+def fetch_last_updated_time(config, last_pull_time):
```

Comment:
this could come as action input from macro

Reply:
Yes. last_pull_time is taken as optional input parameter for the action. In ingestion PB we are passing macro value to this param.